source/common/http/conn_pool_base.cc

```diff
@@ -1,40 +1,292 @@
+      client.codec_client_->numActiveRequests() == 0) {
+    // Close out the draining client if we no longer have active requests.
+    client.codec_client_->close();
+  } else if (client.state_ == ActiveClient::State::BUSY) {
+    // A request was just ended, so we should be below the limit now.
+    ASSERT(client.codec_client_->numActiveRequests() < client.concurrent_request_limit_);
+
+    transitionActiveClientState(client, ActiveClient::State::READY);
+    if (!delay_attaching_request) {
+      onUpstreamReady();
+    }
   }
 }
 
-ConnPoolImplBase::ActiveClient::ConnectionState ConnPoolImplBase::ActiveClient::connectionState() {
-  // We don't track any failure state, as the client should be deferred destroyed once a failure
-  // event is handled.
-  if (connect_timer_) {
-    return Connecting;
+ConnectionPool::Cancellable* ConnPoolImplBase::newStream(Http::StreamDecoder& response_decoder,
+                                                         ConnectionPool::Callbacks& callbacks) {
+  if (!ready_clients_.empty()) {
+    ActiveClient& client = *ready_clients_.front();
+    ENVOY_CONN_LOG(debug, "using existing connection", *client.codec_client_);
+    attachRequestToClient(client, response_decoder, callbacks);
+    return nullptr;
   }
 
-  return Connected;
+  if (host_->cluster().resourceManager(priority_).pendingRequests().canCreate()) {
+    const bool can_create_connection =
+        host_->cluster().resourceManager(priority_).connections().canCreate();
+    if (!can_create_connection) {
+      host_->cluster().stats().upstream_cx_overflow_.inc();
+    }
+
+    // If we have no connections at all, make one no matter what so we don't starve.
+    if ((ready_clients_.empty() && busy_clients_.empty()) || can_create_connection) {
+      createNewConnection();
+    }
+
+    return newPendingRequest(response_decoder, callbacks);
+  } else {
+    ENVOY_LOG(debug, "max pending requests overflow");
+    callbacks.onPoolFailure(ConnectionPool::PoolFailureReason::Overflow, absl::string_view(),
+                            nullptr);
+    host_->cluster().stats().upstream_rq_pending_overflow_.inc();
+    return nullptr;
+  }
+}
+
+void ConnPoolImplBase::onUpstreamReady() {
+  while (!pending_requests_.empty() && !ready_clients_.empty()) {
+    ActiveClientPtr& client = ready_clients_.front();
+    ENVOY_CONN_LOG(debug, "attaching to next request", *client->codec_client_);
+    // Pending requests are pushed onto the front, so pull from the back.
+    attachRequestToClient(*client, pending_requests_.back()->decoder_,
+                          pending_requests_.back()->callbacks_);
+    pending_requests_.pop_back();
+  }
+}
+
+void ConnPoolImplBase::addDrainedCallback(DrainedCb cb) {
+  drained_callbacks_.push_back(cb);
+  checkForDrained();
+}
+
+bool ConnPoolImplBase::hasActiveConnections() const {
+  if (!pending_requests_.empty() || !busy_clients_.empty()) {
+    return true;
+  }
+
+  for (const auto& ready_client : ready_clients_) {
```

Comment:
Do we think this scales sufficiently?  It looks like this is called by freeOnePool which is walking pools looking for something unused, and given I've seen use cases of hundreds to thousands of endpoints and tens of thousands of connections I'm mildly wary of looping if it'll be expensive on busy machines.

Reply:
Yeah good call out. I think we could keep this state actively? At minimum can we TODO this?

Reply:
Good catch.

I think in the common case this shouldn't be too expensive, because it'll stop iterating as soon as it finds a connection with an active request. But there could be cases where this is too expensive.